### PR TITLE
Manual seal mode for dev

### LINKED ISSUE #828: dev mode with instant-seal or without it
dev mode is usefull for local development. 
- [ ] the spec name to be used is open to discussion, maybe all spec ends_with `dev`?
- [ ] also support instant sealing mode.

### PR DESCRIPTION
**Pull request by zqhxuyuan**
## Description

relates to OR closes: #828

- only produce blocks when a transaction is submitted
- add localdev spec config for dev/standalone mode
- extract builder logic into builder.rs
- run single local node example: cargo run -- --chain=dolphin-localdev --alice
- add instant finalized/sealing
image

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari`, `A-dolphin` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [x] This PR is targeted against the *current*  Milestone ( otherwise discuss if it can be added in time)
- [x] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.
- If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated
- Verify benchmarks & weights have been updated for any modified runtime logics

### CODE REVIEW

File: node/src/chain_specs/dolphin.rs [hunk 1]
```diff
@@ -80,13 +80,23 @@ pub fn dolphin_development_config() -> DolphinChainSpec {
 }
 
 /// Returns the Dolphin local chainspec.
-pub fn dolphin_local_config() -> DolphinChainSpec {
+pub fn dolphin_local_config(localdev: bool) -> DolphinChainSpec {
+    let id = if localdev {
+        "calamari_localdev"
```
Comment (ferrell-code) [inline node/src/chain_specs/dolphin.rs]: dolphin_localdev?

File: node/src/chain_specs/dolphin.rs [hunk 2]
```diff
@@ -80,13 +80,23 @@ pub fn dolphin_development_config() -> DolphinChainSpec {
...
+    } else {
+        "calamari_local"
```
Comment (ferrell-code) [inline node/src/chain_specs/dolphin.rs]: same

**Review by ferrell-code — COMMENTED** [review]
Looks good to me

**Review by ferrell-code — DISMISSED** [review]

**Comment by Garandor** [comment]
building from the current commit and running `cargo run -- --chain=dolphin-localdev --alice --force-authoring` i get long log lines of 

```
2023-01-04 14:46:38 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 66 B/s ⬆ 82 B/s    
2023-01-04 14:46:43 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
2023-01-04 14:46:48 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 54 B/s ⬆ 61 B/s    
2023-01-04 14:46:53 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
[...]
2023-01-04 14:49:18 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
2023-01-04 14:49:23 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
2023-01-04 14:49:28 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
2023-01-04 14:49:33 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
```
I gave up after 5min

**Comment by ferrell-code** [comment]
> building from the current commit and running `cargo run -- --chain=dolphin-localdev --alice --force-authoring` i get long log lines of
> 
> ```
> 2023-01-04 14:46:38 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 66 B/s ⬆ 82 B/s    
> 2023-01-04 14:46:43 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> 2023-01-04 14:46:48 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 54 B/s ⬆ 61 B/s    
> 2023-01-04 14:46:53 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> [...]
> 2023-01-04 14:49:18 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> 2023-01-04 14:49:23 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> 2023-01-04 14:49:28 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> 2023-01-04 14:49:33 💤 Idle (0 peers), best: #0 (0x7bd7…bee7), finalized #0 (0x7bd7…bee7), ⬇ 0 ⬆ 0    
> ```
> 
> I gave up after 5min

Try submitting an extrinsic. It won't produce blocks unless a transaction is submitted

We should document this in README

**Comment by Garandor** [comment]
the `--force-authoring` switch seems unnecessary. All it does in nimbus is skip eligibility check of the keys in the node. It'll use the first in store, which when run with `--alice` is the only one anyway

File: node/src/command.rs [hunk 3]
```diff
@@ -411,9 +420,31 @@ pub fn run_with(cli: Cli) -> Result {
...
         None => {
             let runner = cli.create_runner(&cli.run.normalize())?;
+            let chain_spec = &runner.config().chain_spec;
+            let is_dev = chain_spec.is_localdev();
+            info!("id:{}", chain_spec.id());
             let collator_options = cli.run.collator_options();
 
             runner.run_node_until_exit(|config| async move {
+                if is_dev {
+                    info!("DEV STANDALONE MODE.");
```
Comment (Garandor) [inline node/src/command.rs]: Can we give this a big red 🛑 or ⚠️ emoji so it stands out more, the line is a bit buried in the normal logs

File: node/src/builder.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        block_import: client.clone(),
+        parachain_client: client,
+        keystore,
+        skip_prediction: force_authoring,
+        create_inherent_data_providers: provider,
+    }))
+}
+
+/// start standalone dev consensus
+pub fn start_dev_nimbus_instant_seal_consensus<RuntimeApi>(
```
Comment (Garandor) [inline node/src/builder.rs]: shouldnt this be in `service.rs`? It's not a routine that start with `build_`

**Review by Garandor — COMMENTED** [review]
Almost LGTM, looks great with the factored out builder.rs

File: node/src/builder.rs (hunk 4 above)
Comment (zqhxuyuan) [reply node/src/builder.rs]: return `Future` here, and used by `service.rs`.

**Review by Garandor — APPROVED** [review]
Thx for the PR and LGTM

**Review by ferrell-code — APPROVED** [review]

**Review by ghzlatarev — APPROVED** [review]
Nice

**Pull request merged**